prov/verbs/src/verbs_mr.c

```diff
@@ -341,16 +341,54 @@ int fi_ibv_monitor_subscribe(struct ofi_mem_monitor *notifier,
 {
 	struct fi_ibv_domain *domain =
 		container_of(notifier, struct fi_ibv_domain, monitor);
-	int ret;
+	int ret = FI_SUCCESS;
+	RbtStatus rbt_ret;
+	RbtIterator iter;
+	struct iovec *key;
+	struct fi_ibv_subscr_entry *subscr_entry;
+	struct fi_ibv_monitor_entry *entry =
+		calloc(1, sizeof(*entry));
+	if (OFI_UNLIKELY(!entry))
+		return -FI_ENOMEM;
 
 	pthread_mutex_lock(&domain->notifier->lock);
 	ofi_set_mem_free_hook(domain->notifier->prev_free_hook);
 	ofi_set_mem_realloc_hook(domain->notifier->prev_realloc_hook);
 
-	ret = (rbtInsert(domain->notifier->subscr_storage,
-		         (void *)&subscription->iov,
-			 (void *)subscription) != RBT_STATUS_OK) ?
-	       -FI_EAVAIL : FI_SUCCESS;
+	entry->iov = subscription->iov;
+	dlist_init(&entry->subscription_list);
+
+	rbt_ret = rbtInsert(domain->notifier->subscr_storage,
+			    (void *)&entry->iov, (void *)entry);
+	switch (rbt_ret) {
+	case RBT_STATUS_DUPLICATE_KEY:
+		free(entry);
+		iter = rbtFind(domain->notifier->subscr_storage,
+			       (void *)&subscription->iov);
+		assert(iter);
+		rbtKeyValue(domain->notifier->subscr_storage, iter,
+			    (void *)&key, (void *)&entry);
+		/* fall through */
+	case RBT_STATUS_OK:
+		subscr_entry = calloc(1, sizeof(*subscr_entry));
```

Comment:
What happens when calloc() returns NULL here?

Reply:
It returns `-FI_EINVAL` to the caller, `fi_mr_reg` fails in this case.
If it is the first case (not a duplication), it removes the entry from notifier list and free the memory allocated for the notification entry (not a subscription memory for which allocation of memory is unsuccessful)